### PR TITLE
Should interrupt speech synthesis after microphone button is clicked

### LINKED ISSUE #2428: Should interrupt speech synthesis after microphone button is clicked
## Version

`master`

## Describe the bug

This bug is discovered when speech recognition is enabled but speech synthesis is disabled.

Few points:

- On the "core" side, we do not know if speech synthesis is going well or not, or if it is available on the system
- We assumed if speech recognition is available, speech synthesis is also available
   - "core" will mark the next bot message to synthesize, if the user message was sending a message using a microphone
- I believe we should think about whether part of speech should resides on "core", like today. Or we should move everything to "component"
   - "core" has no knowledge about the capability of the system, and should not have
   - Speech is a type of user interface, and "core" is for supporting the UI, should not dictate how a UI would work
   - If we move, we should be able to do the followings:
      - A way to force synthesizing of specific messages (a popular ask)
- When Web Chat is synthesizing, it will not open microphone to prevent echo-ing back into itself
   - We used to stop synthesize when microphone is opened, we have a bug here not stopping synthesis
   - Microphone should have higher priority because microphone is an user interaction, while synthesize may not

## To Reproduce

1. Go to playground, or any speech-enabled Web Chat
2. Click on microphone and say anything that the bot will respond
3. While Web Chat is synthesizing the bot message, click on microphone button immediately

## Expected behavior

Microphone should be turned on immediately, and synthesis is stopped. All messages that are pending synthesis will be stopped as well.

[Bug]

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #2428.

(Will add new tests)

### Fixed

-  Fix [#2428](https://github.com/microsoft/BotFramework-WebChat/issues/2428). Should interrupt speech synthesis after microphone button is clicked, by [@compulim](https://github.com/compulim) in PR [#2429](https://github.com/microsoft/BotFramework-WebChat/pull/2429)

## Description

Today, the microphone button does not stop synthesizing. And since synthesizing is still ongoing, microphone will not turn on to prevent echo.

Tomorrow, the microphone button will stop synthesizing. And this should immediately turn on the microphone.

## Specific Changes

- Add a new saga to stop synthesis when the user want to open the microphone
- Modify speech samples to use authorization token instead of secret

---

-  [x] Testing Added

### CODE REVIEW

File: samples/06.h.cognitive-services-speech-recognition-only/index.html [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    <title>Web Chat: Speech recognition only</title>
+    <!-- Cognitive Services Speech Services adapter is only available in full bundle -->
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+
+    <!--
+      This CDN points to the latest official release of Web Chat. If you need to test against Web Chat's latest bits, please refer to pointing to Web Chat's MyGet feed:
+      https://github.com/microsoft/BotFramework-WebChat#how-to-test-with-web-chats-latest-bits
+    -->
+    <!-- <script src="https://cdn.botframework.com/botframework-webchat/latest/webchat.js"></script> -->
+    <script src="/webchat.js"></script>
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/index.html]: Delete line

File: samples/06.h.cognitive-services-speech-recognition-only/index.html [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    <div id="webchat" role="main"></div>
+    <script>
+      // Create a function to fetch the Cognitive Services Speech Services credentials.
+      // The async function created will hold expiration information about the token and will return cached token when possible.
+      function createFetchSpeechServicesCredentials() {
+        let expireAfter = 0;
+        let lastResult = {};
+
+        return async () => {
+          // Fetch a new token if the existing is expiring.
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/index.html]:
```suggestion
          // Fetch a new token if the existing one is expiring.
```

File: samples/06.h.cognitive-services-speech-recognition-only/index.html [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        });
+
+        const { token } = await directLineTokenRes.json();
+
+        // For this demo, we are using our authorization token. For your production code, you should use a token from your Cognitive Services subscription.
+        const speechServicesTokenRes = await fetch('https://webchat-mockbot.azurewebsites.net/speechservices/token', {
+          method: 'POST'
+        });
+
+        // We are creating a ponyfill factory that will only return the speech recognition part.
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/index.html]:
```suggestion
        // We are creating a ponyfill factory that will only return speech recognition.
```

File: samples/06.h.cognitive-services-speech-recognition-only/index.html [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        async function createSpeechRecognitionOnlyPonyfillFactory() {
+          // Create the ponyfill factory function, that can be called to create a concrete implementation of the ponyfill.
+          const speechServicesPonyfillFactory = await window.WebChat.createCognitiveServicesSpeechServicesPonyfillFactory(
+            {
+              // We are passing the Promise function to the authorizationToken field.
+              // This function will be called every time the token is being used.
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/index.html]:
```suggestion
              // This function will be called every time the token is used.
```

File: samples/06.h.cognitive-services-speech-recognition-only/index.html [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            // We are only returning speech recognition part of the ponyfill, but not the speech synthesis part.
+            return {
+              SpeechGrammarList: speechServicesPonyfill.SpeechGrammarList,
+              SpeechRecognition: speechServicesPonyfill.SpeechRecognition
+            };
+          };
+        }
+
+        // Pass a Web Speech ponyfill factory to renderWebChat.
+        // You can also use your own speech engine, given it is compliant to W3C Web Speech API, https://w3c.github.io/speech-api/.
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/index.html]:
```suggestion
        // You can also use your own speech engine, given it is compliant to W3C Web Speech API: https://w3c.github.io/speech-api/.
```

File: samples/06.h.cognitive-services-speech-recognition-only/README.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
+# Sample - Using speech recognition only
+
+A simple web page that uses speech-to-text only and disable text-to-speech. This sample makes changes that are based off of the [Cognitive Services Speech Services sample](./../06.c.cognitive-services-speech-services-js/README.md).
+
+# Test out the hosted sample
+
+-  [Try out MockBot](https://microsoft.github.io/BotFramework-WebChat/06.f.hybrid-speech)
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/README.md]: Don't forget to update this with the correct link

File: samples/06.h.cognitive-services-speech-recognition-only/README.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+# How to run locally
+
+-  Fork this repository
+-  Navigate to `/Your-Local-WebChat/samples/06.h.cognitive-services-speech-recognition-only` in command line
+-  Run `npx serve`
+-  Browse to [http://localhost:5000/](http://localhost:5000/)
+
+<!--
+# Things to try out
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/README.md]: Why is this all commented out?

File: samples/06.g.select-voice/index.html [hunk 8]
```diff
@@ -39,14 +74,14 @@
         });
         const { token } = await directLineTokenResponse.json();
 
-        const speechTokenResponse = await fetch('https://webchat-mockbot.azurewebsites.net/speechservices/token', {
-          method: 'POST'
-        });
-        const { region, token: authorizationToken } = await speechTokenResponse.json();
+        // Create the ponyfill factory function, that can be called to create a concrete implementation of the ponyfill.
```
Comment (corinagum) [inline samples/06.g.select-voice/index.html]:
```suggestion
        // Create the ponyfill factory function, which can be called to create a concrete implementation of the ponyfill.
```

File: samples/06.g.select-voice/README.md [hunk 9]
```diff
@@ -37,48 +37,26 @@ When Web Chat synthesizes an activity, it will call the `selectVoice` function w
...
   selectVoice: (voices, activity) =>
-    // If the activity is in zh-HK, use a voice with keyword "TracyRUS" (Cantonese).
-    // Otherwise, use "JessaNeural" (preferred) or "Jessa".
-    activity.locale === 'zh-HK' ?
-      voices.find(({ name }) => /TracyRUS/iu.test(name))
-    :
-      voices.find(({ name }) => /JessaNeural/iu.test(name))
+    activity.locale === 'zh-HK'
+    ? voices.find(({ name }) => /TracyRUS/iu.test(name))
+    : voices.find(({ name }) => /JessaNeural/iu.test(name))
```
Comment (corinagum) [inline samples/06.g.select-voice/README.md]: https://github.com/microsoft/BotFramework-WebChat/issues/2424
When making changes here did you notice any problems with selecting a voice? We have a bug reported that voice does not change.

File: samples/06.f.hybrid-speech/index.html [hunk 10]
```diff
@@ -47,9 +82,18 @@
 
         // We are creating a hybrid ponyfill factory that will merge the result of two ponyfill factories together.
         async function createHybridPonyfillFactory({ authorizationToken, region }) {
+          // Create the ponyfill factory function, that can be called to create a concrete implementation of the ponyfill.
```
Comment (corinagum) [inline samples/06.f.hybrid-speech/index.html]:
```suggestion
          // Create the ponyfill factory function, which can be called to create a concrete implementation of the ponyfill.
```

File: samples/06.c.cognitive-services-speech-services-js/index.html [hunk 11]
```diff
@@ -28,36 +28,59 @@
...
-          // In case you are using your own subscription key, please note that client should always authenticate against your server
-          // to avoid exposing the subscription key to any part of your client code.
-          webSpeechPonyfillFactory = await window.WebChat.createCognitiveServicesSpeechServicesPonyfillFactory({
-            region: searchParams.get('r') || 'westus',
-            subscriptionKey
-          });
-        } else {
-          const res = await fetch('https://webchat-mockbot.azurewebsites.net/speechservices/token', { method: 'POST' });
-          const { region, token: authorizationToken } = await res.json();
+        // Create the ponyfill factory function, that can be called to create a concrete implementation of the ponyfill.
```
Comment (corinagum) [inline samples/06.c.cognitive-services-speech-services-js/index.html]:
```suggestion
        // Create the ponyfill factory function, which can be called to create a concrete implementation of the ponyfill.
```

File: samples/06.e.cognitive-services-speech-services-with-lexical-result/index.html [hunk 12]
```diff
@@ -37,29 +72,16 @@
...
-          // to avoid exposing the subscription key to any part of your client code.
-          webSpeechPonyfillFactory = await window.WebChat.createCognitiveServicesSpeechServicesPonyfillFactory({
-            region: searchParams.get('r') || 'westus',
-            subscriptionKey,
-            textNormalization: 'lexical'
-          });
-        } else {
-          const res = await fetch('https://webchat-mockbot.azurewebsites.net/speechservices/token', { method: 'POST' });
-          const { region, token: authorizationToken } = await res.json();
+        // Create the ponyfill factory function, that can be called to create a concrete implementation of the ponyfill.
```
Comment (corinagum) [inline samples/06.e.cognitive-services-speech-services-with-lexical-result/index.html]:
```suggestion
        // Create the ponyfill factory function, which can be called to create a concrete implementation of the ponyfill.
```

File: samples/06.h.cognitive-services-speech-recognition-only/index.html (hunk 2 above)
Comment (compulim) [reply samples/06.h.cognitive-services-speech-recognition-only/index.html]: Will fix all other files.

File: samples/06.g.select-voice/README.md (hunk 9 above)
Comment (compulim) [reply samples/06.g.select-voice/README.md]: Just tested `06.g.select-voice` with `master` build, it works fine.

File: samples/06.h.cognitive-services-speech-recognition-only/README.md (hunk 7 above)
Comment (compulim) [reply samples/06.h.cognitive-services-speech-recognition-only/README.md]: Because, I didn't finish writing it. 🤣

Comment (compulim) [reply samples/06.h.cognitive-services-speech-recognition-only/README.md]: Finished it.

File: samples/06.g.select-voice/index.html [hunk 13]
```diff
@@ -9,7 +9,8 @@
       This CDN points to the latest official release of Web Chat. If you need to test against Web Chat's latest bits, please refer to pointing to Web Chat's MyGet feed:
       https://github.com/microsoft/BotFramework-WebChat#how-to-test-with-web-chats-latest-bits
     -->
-    <script src="https://cdn.botframework.com/botframework-webchat/latest/webchat.js"></script>
+    <!-- <script src="https://cdn.botframework.com/botframework-webchat/latest/webchat.js"></script> -->
+    <script src="/webchat.js"></script>
```
Comment (corinagum) [inline samples/06.g.select-voice/index.html]: Fix this link?

File: samples/06.h.cognitive-services-speech-recognition-only/README.md [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+```diff
+  const speechServicesPonyfillFactory = await window.WebChat.createCognitiveServicesSpeechServicesPonyfillFactory(
+    {
++     authorizationToken: fetchSpeechServicesToken,
++     region: await fetchSpeechServicesRegion()
+    }
+  );
+```
+
+Then, return a new ponyfill factory, that will call the one from Speech Services, but partially returning the result.
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/README.md]:
```suggestion
Then return a new ponyfill factory that will call the one from Speech Services, but partially return the result.
```

File: samples/06.h.cognitive-services-speech-recognition-only/index.html [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+              // In contrast, we are only fetching the region once.
+              region: await fetchSpeechServicesRegion()
+            }
+          );
+
+          return options => {
+            const speechServicesPonyfill = speechServicesPonyfillFactory(options);
+
+            // We are only returning speech recognition part of the ponyfill, but not the speech synthesis part.
```
Comment (corinagum) [inline samples/06.h.cognitive-services-speech-recognition-only/index.html]:
```suggestion
            // We are only returning the speech recognition part of the ponyfill, not the speech synthesis part.
```

**Pull request merged**